t/lib/WTSI/NPG/HTS/TreePublisherTest.pm

```diff
@@ -65,54 +94,14 @@ sub publish_tree : Test(59) {
     return ({attribute => 'extra', value => 'evalue'})
   };
 
-  my $mlwh_json_filename = qq[metadata.json];
-  my $mlwh_json_cb = sub {
-    # DataObject, path to file folder, collection file
-    my ($obj, $collection, $file) = @_;
-      if (defined $mlwh_json_filename) { 
-        Readonly::Scalar my $JSON_FILE_VERSION => '1.0';
-        my $mlwh_hash = {
-          irods_root_collection    => $collection,
-          irods_data_relative_path => $file
-        };
-        my ($json_fh, $json_hash);
-        if (-e $mlwh_json_filename) {
-          open $json_fh, '+<:encoding(UTF-8)', $mlwh_json_filename or
-            self->logcroak(q[could not open ml warehouse json file] .
-            qq[$mlwh_json_filename]);
-          $json_hash = decode_json <$json_fh>;
-        }
-        else {
-          open $json_fh, '>:encoding(UTF-8)', $mlwh_json_filename or
-            self->logcroak(q[could not open ml warehouse json file] .
-            qq[$mlwh_json_filename]);
-          $json_hash = {
-            version  => $JSON_FILE_VERSION,
-            irods_top_collection_folder => $irods_tmp_coll,
-            products => [],
-          };
-        }
-        push @{$json_hash->{products}}, $mlwh_hash;
-
-        seek $json_fh, 0, 0;
-
-        print $json_fh encode_json($json_hash) or
-          self->logcroak(q[could not write to ml warehouse json file ] .
-          qq[$mlwh_json_filename]);
-
-        close $json_fh or
-          self->logcroak(q[could not close ml warehouse json file] .
-          qq[$mlwh_json_filename]);
-      }
-      return 1;
-    };
-
   my ($num_files, $num_processed, $num_errors) =
       $pub->publish_tree(\@files,
                          primary_cb   => $primary_avus,
                          secondary_cb => $secondary_avus,
-                         extra_cb     => $extra_avus,
-                         mlwh_json_cb => $mlwh_json_cb);
+                         extra_cb     => $extra_avus);
+
+  my $mlwh_json_filename = qq[metadata.json];
+  write_json($mlwh_json_filename, $pub->dest_collection);
```

Comment:
You have added this call to the test for `publish_tree`. What are the aiming to test? It's testing a copy of the code from the script, rather than the script.

Reply:
It's good for each tests to focus on a single purpose. This test checks that `publish_tree` in `TreePublisher` performs its basic function i.e. it publishes the expected files with the expected metadata on them.

Your test adds a check on the the function `write_json` which is unrelated to `publish_tree` and additionally doesn't test that `npg_publish_tree.pl` writes the JSON file correctly, but tests a _copy_ of that code that you've added to the tests.